File: ujin/5주차_DP/1912.py
n = int(input())
num = list(map(int, input().split()))
dp = [0] * n
dp[0] = num[0]

for i in range(1, n):
    dp[i] = max(num[i], dp[i-1] + num[i])

print(max(dp))

'''
현재 값과 이전까지의 합 + 현재 값 중 큰 값을 dp에 저장하여 최댓값을 기록하도록 했다.
'''

File: syuzzang/5주차dp/1912.js
let fs = require("fs");
let numbers = fs.readFileSync("./input.txt").toString().split("\n");
const N = Number(numbers[0]);
numbers = numbers[1]
  .trim()
  .split(" ")
  .map((element) => Number(element));
numbers.unshift(0);
const dp = new Array(N + 1);
dp[0] = numbers[1];
dp[1] = numbers[1];

for (let i = 2; i <= N; i++) {
  // 직전까지의 부분 합 + 현재 값  : dp[i-1] + numbers[i]
  // 현재 값 : numbers[i]
  if (dp[i - 1] + numbers[i] >= numbers[i]) {
    // 직전까지의 부분 합 + 현재 값이 가장 클 때
    dp[i] = dp[i - 1] + numbers[i];
  } else {
    // 현재 값이 가장 클 때
    dp[i] = numbers[i];
  }
}
console.log(Math.max(...dp));
